### DISCUSSION TITLE
Adding conditional constraints on slack variables

### DISCUSSION THREAD

**Discussion by jaswanthvishnu10** [google/or-tools#3710] [Routing (and legacy CP) questions]
Hi team,

I am trying to solve a vehicle routing problem where I have 2 types of nodes A & B.

I want to add a constraint such that any vehicle visiting nodes of type B cannot wait for more than X mins. In other words the slack on the time dimension cannot be more than X.

| Path | Slack max |
| ---- | ----------- |
| A->A | INTEGER_MAX
| A->B | X
| B->A | INTEGER_MAX
| B->B | X

Any help on this would be appreciated. Thanks in advance.

**Comment by vuongdanghuy** [comment]
I think you could use timeDimension.SlackVar(index).setValue(X) for every index that is a B node.
See this example of refuel which set positive slack var only for refuel node.
https://github.com/google/or-tools/blob/60fa2fb2719409a9f2dedc38615117012c399e19/examples/cpp/cvrptw_with_refueling.cc#L118-L135

Hope this can help.

**Reply by Mizux** [reply]
Prefer to use `SetRange(0, X)` instead. 
note: `setValue(x)` will "force" the waiting to x while you just want to be "at most" x.

**Reply by jaswanthvishnu10** [reply]
Hi @vuongdanghuy @Mizux , setting range at node B would apply to the slack variables of paths starting from node B right? 
I want to set the range for paths ending at node B. Let me know if there is a mistake in my understanding.
`slack(i) = cumul(j) - cumul(i) - transit(i, j)`

**Reply by vuongdanghuy** [reply]
Oh, i see, your understand is correct.

**Comment by jaswanthvishnu10** [comment]
I have implemented the constraint using the below code. But I am not sure how scalable this implementation is as it involves adding a lot of constraints.
```cpp
for (int i = 0; i < getNodes().size(); i++) {
            Node node = getNodes().get(i);
            if (node.getType().equals("B")) {
                long typeBNodeIndex = manager.nodeToIndex(i);
                for (int j = 0; j < getNodes().size(); j++) {
                    long sourceNodeIndex = manager.nodeToIndex(j);
                    IntExpr isTimeSlackWithinRange = solver.makeConditionalExpression(
                        solver.makeIsEqualVar(model.nextVar(sourceNodeIndex), solver.makeIntConst(typeBNodeIndex)), // Is the B Type node the next node from sourceNode
                        solver.makeIsLessOrEqualVar(timeDimension.slackVar(sourceNodeIndex), solver.makeIntConst(120)), // Max slack is for 2 minutes or 120 seconds
                        1
                    );
                    solver.addConstraint(
                        solver.makeEquality(isTimeSlackWithinRange, 1)
                    );
                }
            }
}
```
Is there any other suggested way to achieve this?

**Comment by Regista6** [comment]
Maybe this could work (py). 
```py 
      for b_node in B:
        next_idx = manager.NodeToIndex(b_node)
        for node in range(num_nodes):
            curr_idx = manager.NodeToIndex(node)
            cond = routing.NextVar(curr_idx) == next_idx
            constraint = time_dim.SlackVar(curr_index).SetRange(0, X)
            expr = routing.solver().ConditionalExpression(cond, constraint, 1)
            routing.solver().Add(expr >= 1)
```

**Reply by jaswanthvishnu10** [reply]
Yeah, my implementation in java is similar to what you've done here in Python.
